### ISSUE TITLE
Saving orignal hashes when XCI > NSP converting with NSCB

### ISSUE THREAD

**Issue by JrRadiant** [julesontheroad/NSC_BUILDER#208]
Hi!

After converting my XCI dump into the .NSP file, and installing it (tried multiple installers), the game cannot pass the system's "Check for Corrupt Data", it constantly shows some errors. I tried many XCI dumps that were converted to NSPs with NSCB, and none of them passed HOS "Check for Corrupt Data" tests, unfortunately.
eshops' dumped NSPs with the tickets are all good, installation and checking are completely OK.

Dear @julesontheroad, can you please save source ncas hashes that were derived from XCI and saving all of them into NSP in a xci>nsp converting process, so NSP that were converted from XCI may successfully pass the "Check for Corrupt Data" in HOS?.. Thank you!

**Comment by duckbill007** [comment]
+1 

Also, xci converted to nsp via 4nxci are all correct and passing "Check for Corrupt Data"

**Comment by julesontheroad** [comment]
NSCB intention is to modify the files with the least changes possible, allowing easy verification and restoration of the original files. 
A hash-patched cnmt can't pass NSCB level2 verification, while it's verification can be postponed to patch the hashes in the cnmt and verify it on the hash check step with the usual rsv and hashes provided the rest of nca files passed the checks i particularly don't like that solution. The verification on other files would not be afected by renaming them though.
Another point that is what actually made 4nxci to add the option to keep hashes is that if you try to install a game, update, etc... that you already have installed with a different set of patches\modification it won't overwrite the original ncas leaving them orphaned, which i now is not a problem for dbi but for other installers.
I'm not against adding an option allowing to patch names or only hashes in then cnmt, i just set it that way intentionally.
Also if you are thinking about mtp xci installation there's another solution to it which is for dbi handling the patching of the nca if you want. Either by checking if they have isgamecard=1 even if are nsp or nsz and then processing as a xci or using other extension for example so it knows it will receive xci files. There's no other solution for installation if converting on the fly since you can't predict the new hashes, in case you think on the console messages. 
**For the cnmt though i could set it at the end and patch it before pushing it, that will make it pass console hash verification, i've been thinking about that these past few days.** It wouldn't prevent the hash missmatch messages while installing for the rest of the ncas though.

**Comment by julesontheroad** [comment]
So basically to abreviate:
- If implemented this option **restore mode** needs a modification to allow restoration if cnmt is the only nca that doesn't pass verification, after that it has to be patched at end of restoration and verified.
- Verification has to be modified to judge to ask to do full hash verification for files where the cnmt is the only file that doesn't pass level2 verification. Something like **nca hashes were probably changed from original source, please do hash verification for full judgement**
Alternatively:
- cnmt can be set to be the final file and patch the hashes before pushing it.

I may be open to these solutions leaving the rehash as an option.

**Comment by duckbill007** [comment]
Sorry, all this iss too complicated for me.
I know following: If I do xci->nsp in 4nxci then it installs without errors and validates through system check without errors.
If I do xci->nsp in NSCB then it gives HASH MISMATCH during install and can not validate through system check.

**Comment by julesontheroad** [comment]
@duckbill007 yeah but you can't restore the nca to original state which was always my aim and since it patches the npdm can't also virtually restore the changes for verification. Anyway as i said i can add a non default option to patch hashes and rename nca, i will never add it as default though since it directly conflicts with the verification and restoration system forcing to go on a roundabout way about it and i won't add acid patching, not for retail files at least.

**Comment by duckbill007** [comment]
If you need some extra info for reverse process, you can store it as txt/xml and add to resulting nsp. All title installers will just skip unknown files and you will get correct files that can be installed without errors and be verified on console without errors.

**Comment by julesontheroad** [comment]
While the idea itself is not bad it feels weird to include xml files on xci since they never include them officially. Of course you can add any kind of file to nsp and xci with extra data though that is lost on processing or even appending the data in other way, squirrel for example has a function to add a footer outside of the declared size of the file making it invisible if you don't know is there which can be used to add extra information.
**The main concern though isn't on original files but on console to pc dumps**, with your suggestion when the file is dump back to pc if you patched something non restorable the original nca state is lost. Originally i thought about adding a db but i prefer to make a solution that doesn't need external information to restore the nca files.
Anyway is reversable as long as you don't patch the acid, just needs to postpone the final assesment till after the file is restored.

**Comment by JrRadiant** [comment]
> I will add that as an option just saying i won't set it as default.

That must be good enough, thank you!
